src/node.cc

```diff
@@ -1303,86 +1303,187 @@ void AddPromiseHook(v8::Isolate* isolate, promise_hook_func fn, void* arg) {
   env->AddPromiseHook(fn, arg);
 }
 
+// Heap-allocatable TryCatch
+namespace {
+struct _TryCatch {
+  explicit _TryCatch(v8::Isolate* isolate) : real_try_catch(isolate) {}
 
-MaybeLocal<Value> MakeCallback(Environment* env,
-                               Local<Value> recv,
-                               const Local<Function> callback,
-                               int argc,
-                               Local<Value> argv[],
-                               async_context asyncContext) {
-  // If you hit this assertion, you forgot to enter the v8::Context first.
-  CHECK_EQ(env->context(), env->isolate()->GetCurrentContext());
+  v8::TryCatch real_try_catch;
+};
+}
 
-  Local<Object> object;
+class InternalCallbackScope {
+ public:
+  InternalCallbackScope(Environment* env,
+                        Local<Object> object,
+                        const async_context& asyncContext);
+  ~InternalCallbackScope();
+  void Close();
 
-  Environment::AsyncCallbackScope callback_scope(env);
-  bool disposed_domain = false;
+  inline bool Failed() const { return failed_; }
+  inline void MarkAsFailed() { failed_ = true; }
+  inline v8::Local<v8::Value> GetFailureReturnValue() const {
+    return failure_return_value_;
+  }
+  inline bool IsInnerMakeCallback() const {
+    return callback_scope_.in_makecallback();
+  }
 
-  if (recv->IsObject()) {
-    object = recv.As<Object>();
+  inline void UseTryCatch() {
+    CHECK_EQ(try_catch_, nullptr);
+    try_catch_ = new _TryCatch(env_->isolate());
+    try_catch_->real_try_catch.SetVerbose(true);
   }
 
+ private:
+  Environment* env_;
+  async_context async_context_;
+  v8::Local<v8::Object> object_;
+  Environment::AsyncCallbackScope callback_scope_;
+  bool failed_ = false;
+  bool pushed_ids_ = false;
+  bool closed_ = false;
+  _TryCatch* try_catch_ = nullptr;
+
+  // This is only for backwards compatibility, semantically it's always empty.
+  v8::Local<v8::Value> failure_return_value_;
+};
+
+CallbackScope::CallbackScope(Isolate* isolate,
+                             Local<Object> object,
+                             async_context asyncContext)
+  : private_(new InternalCallbackScope(Environment::GetCurrent(isolate),
+                                       object,
+                                       asyncContext)) {
+  private_->UseTryCatch();
+}
+
+CallbackScope::~CallbackScope() {
+  delete private_;
+}
+
+InternalCallbackScope::InternalCallbackScope(Environment* env,
+                                             Local<Object> object,
+                                             const async_context& asyncContext)
+  : env_(env),
+    async_context_(asyncContext),
+    callback_scope_(env) {
```

Comment:
this probably stems from a lack of C++ knowledge, but why does passing `Environment*` here properly cast to the `AsyncCallbackScope*` that is `callback_scope_`?

Reply:
> the `AsyncCallbackScope*` that is `callback_scope_`

`callback_scope_` is a `AsyncCallbackScope`, not a pointer – it’s a normal call to its constructor